### PR TITLE
allocator: carve out a package boundary

### PR DESCRIPTION
**Pull request by irfansharif**
**allocator/storepool: carve out package from kvserver**

This commit tries to pave the way for the allocator itself to be carved
out (storepool probably being a direct dependency, but ideally through
an interface). This commit shamelessly exports any and everything needed
in order for StorePool to sit outside the kvserver package boundary.
It's unfortunate how tangled up this component is, including in tests
(some of which reach deep into inner mutexes and mutate state). We're
not doing anything to improve the status quo other than moving it and
maybe pointing to the awkwardness.

**allocatorimpl: carve out an allocator package**

Do the same for kvserver/replicastats, a type that's used as an input
for the allocator at various points (needed to break the dependency
between the allocatorimpl and kvserver). We take the same approach as
the last commit, exporting anything and everything needed to introduce a
hard package boundary.

### CODE REVIEW

File: pkg/kv/kvserver/allocator/base.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	return store.Capacity.FractionUsed() < MaxFractionUsedThreshold
+}
+
+const (
+	// MaxFractionUsedThreshold controls the point at which the store cedes having
+	// room for new replicas. If the fraction used of a store descriptor capacity
+	// is greater than this value, it will never be used as a rebalance or
+	// allocate target and we will actively try to move replicas off of it.
+	MaxFractionUsedThreshold = 0.95
+)
```
Comment (kvoli) [inline pkg/kv/kvserver/allocator/base.go:46]: Why did these move into a separate file? There are other similar constraint checks that should probably also tag along if these do get split here.

nit: the const declaration should be above the func.

File: pkg/kv/kvserver/allocator_scorer.go [hunk 2]
```diff
@@ -46,16 +48,11 @@ const (
 	// https://www.eecs.harvard.edu/~michaelm/postscripts/mythesis.pdf.
 	allocatorRandomCount = 2
 
-	// maxFractionUsedThreshold: if the fraction used of a store descriptor
-	// capacity is greater than this value, it will never be used as a rebalance
-	// or allocate target and we will actively try to move replicas off of it.
-	maxFractionUsedThreshold = 0.95
-
 	// rebalanceToMaxFractionUsedThreshold: if the fraction used of a store
```
Comment (kvoli) [inline pkg/kv/kvserver/allocator_scorer.go:51]: This should also go along to base.go, if you have the other disk fullness check there.

File: pkg/kv/kvserver/allocator_scorer.go [hunk 3]
```diff
@@ -2082,11 +2064,6 @@ func (o storeHealthOptions) rebalanceToReadAmpIsHealthy(
 	return o.enforcementLevel < storeHealthBlockRebalanceTo
 }
 
-// maxCapacityCheck returns true if the store has room for a new replica.
-func maxCapacityCheck(store roachpb.StoreDescriptor) bool {
-	return store.Capacity.FractionUsed() < maxFractionUsedThreshold
-}
-
 // rebalanceToMaxCapacityCheck returns true if the store has enough room to
```
Comment (kvoli) [inline pkg/kv/kvserver/allocator_scorer.go:2081]: Should this also go along to base.go, with the other disk fullness check? I'm don't think there's as much benefit from splitting this off from the scorer yet.

**Review by kvoli — CHANGES_REQUESTED** [review]

File: pkg/kv/kvserver/allocator_scorer.go (hunk 3 above)
Comment (irfansharif) [reply pkg/kv/kvserver/allocator_scorer.go:2081]: See comment above.

File: pkg/kv/kvserver/allocator_scorer.go (hunk 2 above)
Comment (irfansharif) [reply pkg/kv/kvserver/allocator_scorer.go:51]: See comment above.

File: pkg/kv/kvserver/allocator/base.go (hunk 1 above)
Comment (irfansharif) [reply pkg/kv/kvserver/allocator/base.go:46]: Everything that's shared between the subpackages found its way below into this base package, everything that could be scoped to just the subpackage was kept within. MaxCapacityCheck is used by both the allocatorimpl and storepool, which is why only these constants feature here. These boundaries are going to be fuzzy/awkward for now, it's not clear what things should be shared vs. not.

Moved the const declarations above.

**Review by kvoli — APPROVED** [review]

**Comment by irfansharif** [comment]
🍺 

bors r+

**Pull request merged**